// angular2-platform-node/__private_imports__.js
"use strict";
var platform_browser_1 = require('@angular/platform-browser');
var core_1 = require('@angular/core');
var compiler_1 = require('@angular/compiler');
exports.BROWSER_SANITIZATION_PROVIDERS = platform_browser_1.__platform_browser_private__.BROWSER_SANITIZATION_PROVIDERS;
exports.SharedStylesHost = platform_browser_1.__platform_browser_private__.SharedStylesHost;
exports.DomSharedStylesHost = platform_browser_1.__platform_browser_private__.DomSharedStylesHost;
exports.DomRootRenderer = platform_browser_1.__platform_browser_private__.DomRootRenderer;
exports.DomEventsPlugin = platform_browser_1.__platform_browser_private__.DomEventsPlugin;
exports.KeyEventsPlugin = platform_browser_1.__platform_browser_private__.KeyEventsPlugin;
exports.DomAdapter = platform_browser_1.__platform_browser_private__.DomAdapter;
exports.setRootDomAdapter = platform_browser_1.__platform_browser_private__.setRootDomAdapter;
exports.HammerGesturesPlugin = platform_browser_1.__platform_browser_private__.HammerGesturesPlugin;
exports.ViewUtils = core_1.__core_private__.ViewUtils;
exports.AnimationKeyframe = core_1.__core_private__.AnimationKeyframe;
exports.AnimationPlayer = core_1.__core_private__.AnimationPlayer;
exports.AnimationStyles = core_1.__core_private__.AnimationStyles;
exports.RenderDebugInfo = core_1.__core_private__.RenderDebugInfo;
exports.SelectorMatcher = compiler_1.__compiler_private__.SelectorMatcher;
exports.CssSelector = compiler_1.__compiler_private__.CssSelector;
var __empty = null;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = __empty;
